### PR TITLE
TECH: intellij plugin

### PR DESCRIPTION
**Pull request by wilmveel**

### CODE REVIEW

File: lsp/intellij-plugin/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  main {
+    java {
+      srcDirs("src/main/gen")
+    }
+  }
+}
+
+dependencies {
+  implementation(project(":compiler:core"))
+  testImplementation("junit:junit:4.13.2")
```
Comment (jerrevanveluw) [inline lsp/intellij-plugin/build.gradle.kts:30]: junit 4?

File: lsp/intellij-plugin/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  maven{
+    url = uri("https://jitpack.io")
+  }
+}
+
+// Include the generated files in the source set
+sourceSets {
+  main {
+    java {
+      srcDirs("src/main/gen")
```
Comment (jerrevanveluw) [inline lsp/intellij-plugin/build.gradle.kts:23]: not src/main/target/{something} ?

File: lsp/intellij-plugin/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+dependencies {
+  implementation(project(":compiler:core"))
+  testImplementation("junit:junit:4.13.2")
+
+}
+
+java {
+  sourceCompatibility = JavaVersion.VERSION_16
+    targetCompatibility = JavaVersion.VERSION_16
```
Comment (jerrevanveluw) [inline lsp/intellij-plugin/build.gradle.kts:36]: Shouldn't we set this higher for the whole project?

File: lsp/intellij-plugin/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    sinceBuild.set("213")
+    untilBuild.set("222.*")
+  }
+
+  test {
+    // This path value is a machine-specific placeholder text.
+    // Set idea.home.path to the absolute path to the intellij-community source
+    // on your local machine. For real world projects, use variants described in:
+    // https://docs.gradle.org/current/userguide/build_environment.html
+    systemProperty("idea.home.path", "/Users/jhake/Documents/source/comm")
```
Comment (jerrevanveluw) [inline lsp/intellij-plugin/build.gradle.kts:62]: 'works on my machine' code

File: lsp/intellij-plugin/src/main/kotlin/Annotator.kt [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import community.flock.wirespec.compiler.core.tokenize.tokenize
+import community.flock.wirespec.compiler.utils.Logger
+
+
+class Annotator : ExternalAnnotator<List<CompilerException>, List<CompilerException>>() {
+
+    val logger = object : Logger(false) {}
+
+    override fun collectInformation(file: PsiFile): List<CompilerException>? {
+        return WireSpec.tokenize(file.text)
```
Comment (jerrevanveluw) [inline lsp/intellij-plugin/src/main/kotlin/Annotator.kt:21]: make expression body and nonnullable list

File: lsp/intellij-plugin/src/main/kotlin/Annotator.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+            .flatMap { Parser(logger).parse(it) }
+            .fold(
+                ifLeft = { listOf(it) },
+                ifRight = { listOf() }
+            )
+    }
+
+    override fun doAnnotate(collectedInfo: List<CompilerException>): List<CompilerException> {
+        return collectedInfo
```
Comment (jerrevanveluw) [inline lsp/intellij-plugin/src/main/kotlin/Annotator.kt:30]: make expression body

File: lsp/intellij-plugin/src/main/kotlin/FileType.kt [hunk 7]
```diff
@@ -0,0 +1,27 @@
+package community.flock.wire_spec.lsp.intellij_plugin
+
+import javax.swing.Icon
+
+class FileType private constructor() :
+    com.intellij.openapi.fileTypes.LanguageFileType(Language.INSTANCE) {
```
Comment (jerrevanveluw) [inline lsp/intellij-plugin/src/main/kotlin/FileType.kt:6]: import LanguageFileType

File: lsp/intellij-plugin/src/main/kotlin/FileType.kt [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    override fun getName(): String {
+        return "wire-spec"
```
Comment (jerrevanveluw) [inline lsp/intellij-plugin/src/main/kotlin/FileType.kt:8]: these can all be expression bodies

File: lsp/intellij-plugin/src/main/kotlin/Icons.kt [hunk 9]
```diff
@@ -0,0 +1,9 @@
+// Copyright 2000-2022 JetBrains s.r.o. and other contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
+package community.flock.wire_spec.lsp.intellij_plugin
+
+import com.intellij.openapi.util.IconLoader
+import javax.swing.Icon
+
+object Icons {
```
Comment (jerrevanveluw) [inline lsp/intellij-plugin/src/main/kotlin/Icons.kt:7]: if its only used in `FileType` move it there. Maybe even in companion object

File: lsp/intellij-plugin/src/main/kotlin/Language.kt [hunk 10]
```diff
@@ -0,0 +1,9 @@
+// Copyright 2000-2022 JetBrains s.r.o. and other contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
+package community.flock.wire_spec.lsp.intellij_plugin
+
+class Language : com.intellij.lang.Language("wire-spec") {
```
Comment (jerrevanveluw) [inline lsp/intellij-plugin/src/main/kotlin/Language.kt:4]: import alias

File: lsp/intellij-plugin/src/main/kotlin/Lexer.kt [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+            tokens = WireSpec.tokenize(buffer.toString())
+                .fold(
+                    ifLeft = { emptyList() },
+                    ifRight = { it.filterNot { token -> token.type is EndOfProgram } }
+                )
+        }
+    }
+
+    override fun getBufferSequence(): CharSequence {
+        return buffer
```
Comment (jerrevanveluw) [inline lsp/intellij-plugin/src/main/kotlin/Lexer.kt:30]: A lot of methods can be written with expression bodies

File: lsp/intellij-plugin/src/main/kotlin/SyntaxHighlighter.kt [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+            else -> arrayOfNulls(0)
+        }
+
+    override fun getHighlightingLexer(): IntellijLexer {
+        return Lexer()
+    }
+}
+
+class SyntaxHighlighterFactory : SyntaxHighlighterFactory() {
+    override fun getSyntaxHighlighter(project: Project?, virtualFile: VirtualFile?): SyntaxHighlighter {
```
Comment (jerrevanveluw) [inline lsp/intellij-plugin/src/main/kotlin/SyntaxHighlighter.kt:30]: I guess the overwritten `project` and `virtualFile` can be non null

File: lsp/node/server/build.gradle.kts [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        classpath("com.moowork.gradle:gradle-node-plugin:1.2.0")
+    }
+}
+
+plugins {
+    base
+    id("com.moowork.node") version "1.3.1"
+}
+
+tasks.named<NpmTask>("npm_run_build") {
```
Comment (jerrevanveluw) [inline lsp/node/server/build.gradle.kts:23]: Why is NpmTask needed?

File: lsp/node/server/src/index.ts [hunk 14]
```diff
@@ -83,16 +81,32 @@ const documents = new TextDocuments(TextDocument);
 connection.onInitialize(initialize);
 
 documents.onDidChangeContent(async (change) => {
-  await connection.sendDiagnostics({
-    uri: change.document.uri,
-    diagnostics: getDiagnostics(change.document),
-  });
+  connection.console.log("Hello 123 123 123");
```
Comment (jerrevanveluw) [inline lsp/node/server/src/index.ts:84]: should not be merged

**Review by jerrevanveluw — CHANGES_REQUESTED** [review]

File: lsp/intellij-plugin/build.gradle.kts (hunk 1 above)
Comment (wilmveel) [reply lsp/intellij-plugin/build.gradle.kts:30]: Er zit nog geen test in ik verwijder de dep

File: lsp/intellij-plugin/build.gradle.kts (hunk 2 above)
Comment (wilmveel) [reply lsp/intellij-plugin/build.gradle.kts:23]: I will remove this part. Coming from bootstrap project

File: lsp/intellij-plugin/build.gradle.kts (hunk 4 above)
Comment (wilmveel) [reply lsp/intellij-plugin/build.gradle.kts:62]: rm

File: lsp/intellij-plugin/build.gradle.kts (hunk 3 above)
Comment (wilmveel) [reply lsp/intellij-plugin/build.gradle.kts:36]: rm

File: lsp/intellij-plugin/src/main/kotlin/Icons.kt (hunk 9 above)
Comment (wilmveel) [reply lsp/intellij-plugin/src/main/kotlin/Icons.kt:7]: I follow the intellij example here. Like to stay close to it

File: lsp/intellij-plugin/build.gradle.kts [hunk 15]
```diff
@@ -0,0 +1,47 @@
+// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
+
+plugins {
+  id("java")
+  id("org.jetbrains.kotlin.jvm") version "1.7.20-RC"
```
Comment (jerrevanveluw) [inline lsp/intellij-plugin/build.gradle.kts:5]: We probably have a Kotlin Version in `settings.gradle.kts`

File: lsp/intellij-plugin/src/main/kotlin/Annotator.kt [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+import community.flock.wirespec.compiler.core.exceptions.WireSpecException.CompilerException
+import community.flock.wirespec.compiler.core.flatMap
+import community.flock.wirespec.compiler.core.parse.Parser
+import community.flock.wirespec.compiler.core.tokenize.tokenize
+import community.flock.wirespec.compiler.utils.Logger
+
+
+class Annotator : ExternalAnnotator<List<CompilerException>, List<CompilerException>>() {
+
+    val logger = object : Logger(false) {}
```
Comment (jerrevanveluw) [inline lsp/intellij-plugin/src/main/kotlin/Annotator.kt:18]: private

**Review by jerrevanveluw — APPROVED** [review]

**Pull request merged**